controllers/azurejson_machinetemplate_controller.go

```diff
@@ -56,12 +59,34 @@ func (r *AzureJSONTemplateReconciler) SetupWithManager(ctx context.Context, mgr
 	)
 	defer done()
 
-	return ctrl.NewControllerManagedBy(mgr).
+	azureMachineTemplateMapper, err := util.ClusterToObjectsMapper(r.Client, &infrav1.AzureMachineTemplateList{}, mgr.GetScheme())
+	if err != nil {
+		return errors.Wrap(err, "failed to create mapper for Cluster to AzureMachines")
+	}
+
+	c, err := ctrl.NewControllerManagedBy(mgr).
 		WithOptions(options).
 		For(&infrav1.AzureMachineTemplate{}).
 		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(log, r.WatchFilterValue)).
 		Owns(&corev1.Secret{}).
-		Complete(r)
+		Build(r)
+
+	if err != nil {
+		return errors.Wrap(err, "failed to create controller")
+	}
+
+	// Add a watch on Clusters to requeue when the infraRef is set. This is needed because the infraRef is not initially
+	// set in Clusters created from a ClusterClass.
+	if err := c.Watch(
+		&source.Kind{Type: &clusterv1.Cluster{}},
```

Comment:
@CecileRobertMichon It looks like this fixed the issue. Do we want to be a bit more precise and add a condition to the Watch to update only when the infrastructureRef changes?

Reply:
Do you mean a predicate? Yes that would be good to filter requests so we only queue Clusters that have infra Ref set

Reply:
What predicate should we use here? The only one I can seem to find is `predicates.ClusterUnpausedAndInfrastructureReady(log)` but that refers to the InfrastructureReady status instead of it being set.

Reply:
we could add a new predicate

You're correct that we don't want to use `ClusterUnpausedAndInfrastructureReady` since we want to reconcile secrets before waiting for the cluster to be ready